### PR TITLE
Deprecate ScalarValue bitor, bitand, and bitxor (#6842)

### PR DESCRIPTION
**Pull request by tustvold**
## Which issue does this PR close?

Part of #6842

### CODE REVIEW

File: datafusion/physical-expr/src/aggregate/bit_and_or_xor.rs [hunk 1]
```diff
@@ -678,34 +611,39 @@ impl PartialEq<dyn Any> for DistinctBitXor {
...
-            self.hash_values
+            let values = self
+                .values
                 .iter()
-                .for_each(|distinct_value| distinct_values.push(distinct_value.clone()));
-            vec![ScalarValue::new_list(
-                Some(distinct_values),
-                self.data_type.clone(),
-            )]
+                .map(|x| ScalarValue::new_primitive::<T>(Some(*x), &T::DATA_TYPE))
```
Comment (tustvold) [inline datafusion/physical-expr/src/aggregate/bit_and_or_xor.rs:643]: This is pretty gross, it occurs to me that ScalarValue not storing ArrayRef is pretty strange

Comment (tustvold) [reply datafusion/physical-expr/src/aggregate/bit_and_or_xor.rs:643]: Created https://github.com/apache/arrow-datafusion/issues/7352

Comment (Dandandan) [reply datafusion/physical-expr/src/aggregate/bit_and_or_xor.rs:643]: 👍  I think this came up in a discussion before (distinct count I think).

File: datafusion/common/src/scalar.rs [hunk 2]
```diff
@@ -1781,6 +1782,25 @@ macro_rules! eq_array_primitive {
 }
 
 impl ScalarValue {
+    /// Create a [`ScalarValue`] with the provided value and datatype
+    ///
+    /// # Panics
+    ///
+    /// Panics if d is not compatible with T
+    pub fn new_primitive<T: ArrowPrimitiveType>(
```
Comment (tustvold) [inline datafusion/common/src/scalar.rs:1790]: This is the sort of thing that becomes so much simpler with #7353

**Review by alamb — APPROVED** [review]
LGTM -- thank you @tustvold 

I also think this likely will make the various bit aggregates substantially faster as well now that they have type specialized implementations rather than dynamic dispatch

cc @izveigor

**Review by Dandandan — APPROVED** [review]

**Pull request merged**